### PR TITLE
Advanced training

### PR DESCRIPTION
**Pull request by shilpan97**

### CODE REVIEW

**Comment by enochtsang** [comment]
Hey when did you start this? Could be record time.

**Comment by shilpan97** [comment]
Started to work on this less than 2 weeks ago.

File: Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 1]
```diff
@@ -1,13 +1,18 @@
 #include "BatteryStateOfChargeService.h"
+#include "BatteryData.h"
 
 namespace
 {
     const double BATTERY_AMP_HOUR_CAPACITY = 123.0;
+    const double HOUR_TO_MILLISECONDS = 36000000.0;
+
```
Comment (safasyed) [inline Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: extra whitespace

File: Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 2]
```diff
@@ -16,23 +21,45 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
...
+    QTime time(0,0,0,0);
+    int time_left;
+    if(isCharging()){
+        time_left = (totalAmpHoursUsed()/ current_t) * HOUR_TO_MILLISECONDS;
+    }
+    else
+    {
+        time_left = ((BATTERY_AMP_HOUR_CAPACITY - totalAmpHoursUsed())/current_t) * HOUR_TO_MILLISECONDS;
+    }
+
```
Comment (safasyed) [inline Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: extra whitespace

File: Advanced-Recruit-Training/src/BatteryStateOfChargeService.h [hunk 3]
```diff
@@ -25,4 +25,11 @@ class BatteryStateOfChargeService : public I_BatteryStateOfChargeService
     const double initialStateOfChargePercent_;
 
     // Add your necessary variables here
+    double current_t;
+    double voltage_t;
+    QTime time_t;
+    double prevCurr;
+    QTime prevTime;
+    double AmpHrUsed;
+
```
Comment (safasyed) [inline Advanced-Recruit-Training/src/BatteryStateOfChargeService.h]: extra whitespace

File: Advanced-Recruit-Training/src/LogFileReader.cpp [hunk 4]
```diff
@@ -62,10 +62,28 @@ bool LogFileReader::parseLine(const QString& line, BatteryData& batteryData) con
     QString timeString = sections.at(0);
     batteryData.time = QTime::fromString(timeString, STRING_TIME_FORMAT);
 
-    batteryData.voltage = sections.at(1).toDouble();
+    if(sections.length() < COLUMNS || sections.length() > COLUMNS)
```
Comment (safasyed) [inline Advanced-Recruit-Training/src/LogFileReader.cpp]: You're essentially saying if the length of the section isn't COLUMNS return false so you can write this in a shorter if statement

File: Advanced-Recruit-Training/src/LogFileReader.cpp [hunk 5]
```diff
@@ -62,10 +62,28 @@ bool LogFileReader::parseLine(const QString& line, BatteryData& batteryData) con
...
+        qDebug()<<"DoubleConversion is unsuccessful";
+        return false;
+    }
+
+    if(batteryData.time.isValid())
+        return true;
+    else{
+        qDebug()<<"time is invalid";
+        return false;
+    }
```
Comment (safasyed) [inline Advanced-Recruit-Training/src/LogFileReader.cpp]: extra whitespace

File: Advanced-Recruit-Training/src/LogFileReader.cpp [hunk 6]
```diff
@@ -62,10 +62,28 @@ bool LogFileReader::parseLine(const QString& line, BatteryData& batteryData) con
...
+    bool DoubleConversion1;
+    bool DoubleConversion2;
 
-    return true;
+    batteryData.voltage = sections.at(1).toDouble(&DoubleConversion1);
+
+    batteryData.current = sections.at(2).toDouble(&DoubleConversion2);
+
+    if(DoubleConversion1 != true || DoubleConversion2 != true)
+        return false;
```
Comment (safasyed) [inline Advanced-Recruit-Training/src/LogFileReader.cpp]: Put curly brackets for all if statements even if there is only one line of code in the statement

File: Advanced-Recruit-Training/src/LogFileReader.cpp [hunk 7]
```diff
@@ -62,10 +62,28 @@ bool LogFileReader::parseLine(const QString& line, BatteryData& batteryData) con
...
+    else{
+        qDebug()<<"DoubleConversion is unsuccessful";
+        return false;
+    }
+
+    if(batteryData.time.isValid())
+        return true;
```
Comment (safasyed) [inline Advanced-Recruit-Training/src/LogFileReader.cpp]: curly brackets, same as above

File: Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 8]
```diff
@@ -16,23 +21,45 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
...
+    return time.addMSecs(time_left);
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
     Q_UNUSED(batteryData);
```
Comment (safasyed) [inline Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: remove this

File: Advanced-Recruit-Training/src/BatteryStateOfChargeService.h [hunk 9]
```diff
@@ -25,4 +25,11 @@ class BatteryStateOfChargeService : public I_BatteryStateOfChargeService
     const double initialStateOfChargePercent_;
 
     // Add your necessary variables here
+    double current_t;
+    double voltage_t;
+    QTime time_t;
+    double prevCurr;
+    QTime prevTime;
+    double AmpHrUsed;
```
Comment (bill-luu) [inline Advanced-Recruit-Training/src/BatteryStateOfChargeService.h]: Use `camelCase` in variables.
For private variables, use trailing underscores.

File: Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 10]
```diff
@@ -1,13 +1,18 @@
...
 {
     const double BATTERY_AMP_HOUR_CAPACITY = 123.0;
+    const double HOUR_TO_MILLISECONDS = 36000000.0;
+
 }
 
 BatteryStateOfChargeService::BatteryStateOfChargeService(double initialStateOfChargePercent)
 : initialStateOfChargePercent_(initialStateOfChargePercent)
 {
+    AmpHrUsed = BATTERY_AMP_HOUR_CAPACITY - ( BATTERY_AMP_HOUR_CAPACITY * (initialStateOfChargePercent /100));
```
Comment (bill-luu) [inline Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: Space before and after operators.

File: Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 11]
```diff
@@ -16,23 +21,45 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
...
-    // i.e: Taking a running average of your data values, using most recent data points, etc.
+
+    current_t = batteryData.current;
+    voltage_t = batteryData.voltage;
+    time_t = batteryData.time;
+
+    prevCurr = current_t;
+    prevTime = time_t;
+
+    double avgCurr = (prevCurr + current_t) / 2 ;
```
Comment (bill-luu) [inline Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: I would do 2.0 instead. 
Extra whitespace between 2 and semicolon.

File: Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 12]
```diff
@@ -16,23 +21,45 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
...
     Q_UNUSED(batteryData);
-    // This is where you can update your variables
-    // Hint: There are many different ways that the totalAmpHoursUsed can be updated
-    // i.e: Taking a running average of your data values, using most recent data points, etc.
+
+    current_t = batteryData.current;
+    voltage_t = batteryData.voltage;
+    time_t = batteryData.time;
+
+    prevCurr = current_t;
```
Comment (bill-luu) [inline Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: Remaining `current_t` to something like `presentCurrent` would go better with `previousCurrent`

File: Advanced-Recruit-Training/src/LogFileReader.cpp (hunk 4 above)
Comment (bill-luu) [reply Advanced-Recruit-Training/src/LogFileReader.cpp]: This is basically saying if sections.length() does not equal columns!

File: Advanced-Recruit-Training/src/BatteryStateDisplayService.cpp [hunk 13]
```diff
@@ -27,7 +27,6 @@ void BatteryStateDisplayService::handleBatteryDataReceived(const BatteryData& ba
 
     QTextStream(stdout) << "Voltage: " << batteryData.voltage
         << " Current: " << batteryData.current
-        << " Total Ah used: " << batteryStateOfChargeService_.totalAmpHoursUsed() << endl;
-
-    // TODO: Print out time until the battery is fully charged or depleted.
+        << " Total Ah used: " << batteryStateOfChargeService_.totalAmpHoursUsed()
+        << " Time Remaining: " << batteryStateOfChargeService_.timeWhenChargedOrDepleted().toString("hh:mm:ss.zz") <<endl;
```
Comment (bill-luu) [inline Advanced-Recruit-Training/src/BatteryStateDisplayService.cpp]: This function should tell us the time remaining _until the battery is charged or depleted_, just having the time doesn't tell us much.

**Comment by shilpan97** [comment]
should be good now!

**Pull request closed without merging**